cvmfs/cache_ram.cc

```diff
@@ -0,0 +1,250 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+#include "cvmfs_config.h"
+#include "cache_ram.h"
+
+#include <errno.h>
+#include <assert.h>
+#include <new>
+#include <string.h>
+
+#include "kvstore.h"
+#include "smalloc.h"
+#include "util/posix.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace cache {
+
+
+int RamCacheManager::AddFd(const ReadOnlyFd &fd) {
+  unsigned i = 0;
+  for ( ; i < open_fds_.size(); ++i) {
+    if (open_fds_[i].handle == invalid_fd_) {
+      open_fds_[i] = fd;
+      return i;
+    }
+  }
+  open_fds_.push_back(fd);
+  return i;
+}
+
+bool RamCacheManager::AcquireQuotaManager(QuotaManager *quota_mgr) {
+  return false;
+}
+
+
+int RamCacheManager::Open(const shash::Any &id) {
+  WriteLockGuard guard(rwlock_);
+  return DoOpen(id);
+}
+
+int RamCacheManager::DoOpen(const shash::Any &id) {
+  kvstore::MemoryBuffer buf;
+  ReadOnlyFd fd(id, 0);
+
+  if (pinned_entries_.Ref(id)) {
+    return AddFd(fd);
+  } else if (regular_entries_.PopBuffer(id, &buf) ||
+             volatile_entries_.PopBuffer(id, &buf)) {
+    pinned_entries_.Commit(id, buf);
+    assert(pinned_entries_.Ref(id));
+    return AddFd(fd);
+  } else {
+    return -ENOENT;
+  }
+}
+
+
+int64_t RamCacheManager::GetSize(int fd) {
+  ReadLockGuard guard(rwlock_);
+  if (!IsValid(fd)) return -EBADFD;
+  return pinned_entries_.GetSize(open_fds_[fd].handle);
+}
+
+
+int RamCacheManager::Close(int fd) {
+  kvstore::MemoryBuffer buf;
+  ReadOnlyFd file_descriptor;
+  bool sweep_tail = false;
+  {
+    ReadLockGuard guard(rwlock_);
+    if (!IsValid(fd)) return -EBADFD;
+    file_descriptor = open_fds_[fd];
+    assert(pinned_entries_.GetBuffer(file_descriptor.handle, &buf));
+  }
+
+  WriteLockGuard guard(rwlock_);
+  assert(pinned_entries_.Unref(file_descriptor.handle));
```

Comment:
`assert()` might be optimized away (it isn't in cvmfs, but in general it might).  So it's better to store the result of `Unref` in a variable and test the variable in assert.